model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java

```diff
@@ -88,15 +94,21 @@ public long delete(ModelCriteriaBuilder<M> criteria) {
         }
         Predicate<? super K> keyFilter = b.getKeyFilter();
         Predicate<? super V> entityFilter = b.getEntityFilter();
-        res = 0;
-        for (Iterator<Entry<K, V>> iterator = store.entrySet().iterator(); iterator.hasNext();) {
-            Entry<K, V> next = iterator.next();
-            if (keyFilter.test(next.getKey()) && entityFilter.test(next.getValue())) {
-                res++;
-                iterator.remove();
-            }
+        Stream<Entry<K, V>> storeStream = store.entrySet().stream();
+        final AtomicLong res = new AtomicLong(0);
+
+        if (!queryParameters.getOrderBy().isEmpty()) {
+            Comparator<V> comparator = MapFieldPredicates.getComparator(queryParameters.getOrderBy().stream());
+            storeStream = storeStream.sorted((entry1, entry2) -> comparator.compare(entry1.getValue(), entry2.getValue()));
         }
-        return res;
+
+        paginatedStream(storeStream.filter(next -> keyFilter.test(next.getKey()) && entityFilter.test(next.getValue()))
+                , queryParameters.getOffset(), queryParameters.getLimit())
```

Comment:
@hmlnarik Does it make sense to have `limit` set also for delete method? This can be causing some problems in the future when we, for example, remove a realm with many clients and only 100 (default limit set in QueryParameters) is removed. Maybe we should ignore pagination for deletes completely as currently we don't manually set it anywhere in the code, and default limit 100 may cause problems. WDYT?

Reply:
@hmlnarik Bad news, I am not sure currently, we can set the default limit for each query to 100. There were some test failures with it. The problem is, that we do something like this:
```
policyStore.findByResourceServer(id).stream()
            .map(Policy::getId)
            .forEach(policyStore::delete);
```

This code is executed when you remove a resource server (for example when disabling authz for a client). If the client has more than 100 policies, there are some policies left in the database. This can be fixed however, I would again do it as a follow up enhancement including setting some default value for the `limit` parameter. WDYT?

Reply:
The problem is, that in this case, we are reading from the storage and then remove each policy we found. We would need to find all places like this and replace them with bulk deletes. But I am not sure it will be possible in each case as it would maybe require some join queries to achieve this.

Reply:
Looking at this for the second time: I would argue that this is a misuse of reading - there should be a method "delete all policies belonging to given resource server" in the storage, e.g. `policyStore.deleteByResourceServer(id)`. This way, even caching would be able to invalidate the right events.
I get your point though. How many places like this are in the codebase? It seems that they are located to fine-grained permissions?

Reply:
This is a good question. I don't know the answer yet. Potentially a lot, maybe only a few. I quickly looked at some occurences and clients/roles suffer from the same thing, so this is not only authz:
https://github.com/keycloak/keycloak/blob/a8fdd79d1b9d78e94a7025e2a3f367561fd7557d/model/map/src/main/java/org/keycloak/models/map/client/MapClientProvider.java#L177-L184

I would suggest to merge this as is, without the default value 100 for limit (current state of this PR), and add all discussed things in this PR in a separate Jira. The advantage would be, that all work that follows on this (Jpa, HotRod) can use changes from this PR immediately and these enhancements can be added anytime later as there won't be many conflicts probably. WDYT?

The other option is to address it in this PR, however I am not sure how much time I will be able to spend on this as I need to investigate the HotRod stuff.